cluster-api/CONTRIBUTING.md

```diff
@@ -0,0 +1,60 @@
+# Cluster API Development Guide
+
+## Cloud Provider Dev Guide
+
+### Overview
+
+The Cluster API is a Kubernetes project to bring declarative, Kubernetes-style APIs to cluster creation, configuration, and management. It provides optional, additive functionality on top of core Kubernetes.
+
+This document is meant to help OSS contributors implement support for cloud providers.
+
+As part of adding support for a provider (cloud or on-prem), you will need to:
+
+1.  Create tooling that conforms to the Cluster API (described further below)
+1.  A machine controller that can run independent of the cluster. This controller should handle the lifecycle of the machines, whether it's run inside the master in a pod, in an external API server, or even on your local workstation.
+
+### Resources
+
+*   [kubernetes/kube-deploy](https://github.com/kubernetes/kube-deploy)
+*   [Cluster Management API KEP](https://github.com/kubernetes/community/blob/master/keps/sig-cluster-lifecycle/0003-cluster-api.md)
+*   [Cluster type](https://github.com/kubernetes/kube-deploy/blob/fafc50e6420783179fd5fc7d73c7453f1de68eb4/cluster-api/api/cluster/v1alpha1/types.go#L32)
+*   [Machine type](https://github.com/kubernetes/kube-deploy/blob/fafc50e6420783179fd5fc7d73c7453f1de68eb4/cluster-api/api/cluster/v1alpha1/types.go#L161)
+
+[TODO: Add architecture diagram](https://github.com/kubernetes/kube-deploy/issues/546)
+
+### A new Node can be created in a declarative way
+
+**A new Node can be created in a declarative way, including Kubernetes version and container runtime version. It should also be able to specify provider-specific information such as OS image, instance type, disk configuration, etc., though this will not be portable.**
+
+When a cluster is first created with a cluster config file, there is no master node or api server. So the user will need to bootstrap a cluster. While the implementation details are specific to the provider, the following guidance should help you:
+
+*   Your tooling should create a Node for the master machine specified by the user. This master can use kubeadm to initialize the master. See [this example](https://github.com/kubernetes/kube-deploy/blob/68e27e43894efebb45f5f014aa5510c11015c3b3/cluster-api-gcp/cloud/google/templates.go).
+*   Start the machine controller specific to the provider.
```

Comment:
These will change with the extension apiserver. The pattern will be more like spin up the external apiserver and machine controller, create objects, pivot the apiserver and machine controller into the cluster. Not sure if you want to put in steps that will get changed pretty soon.

Reply:
As we are/will going through big change like Jessica outlined, I would vote to not document old behavior here as it will introduce more confusing.

Reply:
That's a good point. I've changed these steps.